Added a progress bar to the control panel

chandrashekar-s opened the pull request:
- Add Progress bar to the control panel

## Description of what I changed

This PR contains the changes for the issue #449 . The following changes are made.

- Progress bar is added to the control panel.
- Currently, the progress bar is enabled for FlinkRunner, if any other is used then the default spinning GIF will be used by the control panel.
- Flink Joblisteners (org.apache.flink.core.execution.JobListener) are registered during the pipeline job submission which are later used to retrieve the run time metrics of the pipeline.

![image](https://user-images.githubusercontent.com/29541718/229801094-ad2894b2-21fc-4008-84d1-8f3aee0ed53b.png)

<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->
<!--- If your PR is related to an issue , please mention it here. -->
<!--- It is generally a good practice to first file an issue with enough
  context and reference it in the PR, but if you don't have that, please remove
  this section. -->

## E2E test

<!-- There are different scenarios for using the tools in this repo; please
  help your reviewers by describing how you have e2e tested your change. -->

TESTED:

Please replace this with a description of how you tested your PR beyond the
automated e2e/unit tests.

## Checklist: I completed these to help reviewers :)

<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->

- [x] I have read and will follow the [review process](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/master/doc/review_process.md).
- [x] I am familiar with Google Style Guides for the language I have coded in.

  No? Please take some time and review [Java](https://google.github.io/styleguide/javaguide.html) and [Python](https://google.github.io/styleguide/pyguide.html) style guides.

- [x] My IDE is configured to follow the Google [**code styles**](https://google.github.io/styleguide/).

  No? Unsure? -> [configure your IDE](https://github.com/google/google-java-format).

- [ ] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)
- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
- [x] All new and existing **tests passed**.
- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

bashir2 commented on pipelines/batch/src/main/java/org/openmrs/analytics/FhirEtl.java:
```diff
@@ -285,7 +300,13 @@ static Pipeline buildHapiJdbcFetch(
       }
+
       foundResource = true;
+      PCollection<KV<String, Long>> initialPCollection =
```
nit: Please add a TODO in `runFhirFetch` to add this in that function too such that this approach works when we use FHIR-search mode.

chandrashekar-s replied:
Done

bashir2 commented on pipelines/batch/src/main/java/org/openmrs/analytics/FhirEtl.java:
```diff
@@ -285,7 +300,13 @@ static Pipeline buildHapiJdbcFetch(
+          pipeline.apply(
+              "Metric parameters for " + resourceType,
+              Create.of(KV.of(resourceType, Long.valueOf(numResources))));
+      initialPCollection.apply(ParDo.of((new LogAsMetric())));
```
This is an interesting way of injecting these values into the `MetricResults` of the pipeline but is there no easier way of achieving this without a `DoFn`? I mean these form a small set of resourceType->number maps which can be passed to any `DoFn`, e.g., as constructor parameter, would that option make this easier to achieve?

bashir2 replied:
If you can address it before merge it would be nice. I don't mean to actually do what I suggested but analyze that approach and make a decision whether to go with this or change it to that (i.e., remove the TODO in either case).

chandrashekar-s replied:
The metrics could not be logged when injected via the constructor of the beginning pipeline stage i.e. in the JdbcFetchHapi.generateQueryParameters() method (looks like the pipeline context is not available at this stage). Hence retaining the current implementation.

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/metrics/StatsHelper.java:
```diff
@@ -0,0 +1,54 @@
+public class StatsHelper {
+  private static final Logger logger = LoggerFactory.getLogger(StatsHelper.class.getName());
+
+  public static Stats createStats(MetricQueryResults metricQueryResults) {
```
nit: I don't mind if you move this static method to `Stats` class and remove `StatsHelper` to have less classes; but up to you. Please consider adding javadoc for this public method too, especially documenting how a `null` argument is handled. I would also go an additional step and annotate the argument as `@Nullable`.

chandrashekar-s replied:
Refactored and added javadoc

bashir2 replied:
Great, thanks.

bashir2 commented on pipelines/controller/src/main/resources/templates/index.html:
```diff
@@ -106,11 +131,15 @@
-          A pipeline is currently running!
+        <div class="alert alert-primary alert-dismissible fade show" role="alert" th:if="${isRunning}">
+          <div class="progress" >
+            <div class="progress-bar progress-bar-striped" style="min-width: 30px; display: none"></div>
```
Consider adding a snapshot of this while the pipeline is running, to the PR description.

chandrashekar-s replied:
Done

bashir2 commented on pipelines/controller/src/main/resources/templates/index.html:
```diff
@@ -59,19 +78,25 @@
+              showPipelineProgress(stats)
             }
-            setTimeout(function() {
+            setTimeout(function () {
```
nit: remove the additional space after `function` if it is not needed.

chandrashekar-s replied:
Done

bashir2 reviewed (commented):
Thanks @chandrashekar-s for figuring this out and the nice implementation. Most of my comments are minor or questions.

chandrashekar-s commented on pipelines/batch/src/main/java/org/openmrs/analytics/FhirEtl.java:
```diff
@@ -258,6 +260,19 @@ private static void validateOptions(FhirEtlOptions options)
+  /** A simple DoFn that captures the gauge metric of the given input type */
+  public static class LogAsMetric extends DoFn<KV<String, Long>, Void> {
+    @ProcessElement
+    public void processElement(ProcessContext c) {
```
Done

chandrashekar-s commented on pipelines/batch/src/main/java/org/openmrs/analytics/ParquetMerger.java:
```diff
@@ -124,8 +124,9 @@ static Pipeline createMergerPipeline(ParquetMergerOptions options, FhirContext f
-    Counter numOutputRecords = Metrics.counter(EtlUtils.METRICS_NAMESPACE, "numOutputRecords");
-    Counter numDuplicates = Metrics.counter(EtlUtils.METRICS_NAMESPACE, "numDuplicates");
+    Counter numOutputRecords =
+        Metrics.counter(MetricsConstants.METRICS_NAMESPACE, "numOutputRecords");
```
Done

chandrashekar-s commented on pipelines/controller/src/main/java/org/openmrs/analytics/metrics/FlinkPipelineMetrics.java:
```diff
@@ -0,0 +1,78 @@
+import org.springframework.stereotype.Component;
+
+@Component
+@Scope("singleton")
```
Done

chandrashekar-s commented on pipelines/controller/src/main/java/org/openmrs/analytics/ApiController.java:
```diff
@@ -49,11 +54,25 @@ public String runBatch(@RequestParam(name = "isFullRun", required = true) boolea
+    try {
+      metricQueryResults = pipelineManager.getPipelineMetrics();
+    } catch (MetricsNotSupportedException e) {
+      logger.warn("Metrics currently not supported for the pipeline.", e);
```
Done

chandrashekar-s commented on pipelines/controller/src/main/java/org/openmrs/analytics/metrics/PipelineMetricsFactory.java:
```diff
@@ -0,0 +1,43 @@
+      case FLINK_RUNNER:
+        return flinkPipelineMetrics;
+      default:
+        throw new MetricsNotSupportedException(
```
Refactored the method to return null instead of an exception if the metrics is not supported, updated javadoc as well.

Yes, if the runner is not FlinkRunner, then it is defaulted to current behaviour where in a loading icon appears instead of the progress bar. To support additional runners, we need to implement the PipelineMetrics interface to return the MetricQueryResults object correctly.

chandrashekar-s commented:
Thanks Bashir for the review, I have addressed the comments and checked in the new changes. Please have a look.
Apart from the review changes, the below additional changes have been made in the latest commit
 - Captured the mapped resources metric (number of resources fetched and mapped from the FHIR server).
 - Stats.createStats() method is updated to calculate the completion ratio using the metrics from multiple steps in the pipeline

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/PipelineManager.java:
```diff
@@ -243,6 +251,7 @@ synchronized void runBatchPipeline() throws IOException, PropertyVetoException,
     Preconditions.checkState(!isRunning(), "cannot start a pipeline while another one is running");
     PipelineConfig pipelineConfig = dataProperties.createBatchOptions();
     FhirEtlOptions options = pipelineConfig.getFhirEtlOptions();
+    options.setBatchRun(true);
```
It seems we are using this new flag just for storing the pipeline mode and later accessing it, right? So for example it doesn't make sense for a direct user of the pipeline to set this. In that case, I think it makes more sense to move this to a simple class attribute in `PipelineThread`. So when you create an instance of it in `currentPipeline` below, you simply pass true/false for this flag and we don't need to leak it into `FhirEtlOptions`.

chandrashekar-s replied:
Agreed that the batchRun flag need not be exposed via pipeline option class, refactored the code to store the value as a class variable in the PipelineThread class.

bashir2 commented on pipelines/controller/src/test/java/org/openmrs/analytics/metrics/StatsTest.java:
```diff
@@ -0,0 +1,90 @@
+import org.junit.jupiter.params.provider.MethodSource;
+import org.openmrs.analytics.MetricsConstants;
+
+public class StatsTest {
```
Thanks for adding these unit-tests.

bashir2 reviewed (commented):
Thanks for the changes and sorry for the delay.

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/PipelineManager.java:
```diff
@@ -251,13 +250,13 @@ synchronized void runBatchPipeline() throws IOException, PropertyVetoException,
     } else {
-      currentPipeline = new PipelineThread(pipeline, this, dataProperties, pipelineConfig);
+      currentPipeline =
+          new PipelineThread(pipeline, this, dataProperties, pipelineConfig, Boolean.TRUE);
```
nit: why not just `true`?

chandrashekar-s replied:
done

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/PipelineManager.java:
```diff
@@ -300,7 +299,8 @@ synchronized void runIncrementalPipeline()
       currentPipeline =
-          new PipelineThread(pipeline, mergerOptions, this, dataProperties, pipelineConfig);
+          new PipelineThread(
+              pipeline, mergerOptions, this, dataProperties, pipelineConfig, Boolean.FALSE);
```
ditto

chandrashekar-s replied:
done

bashir2 reviewed (approved):
Thanks @chandrashekar-s please feel free to merge after addressing the minor suggestions.

chandrashekar-s commented:
Thanks Bashir for the comments, few changes have been made to resolve them. Please have a look.

bashir2 commented:
> Thanks Bashir for the comments, few changes have been made to resolve them. Please have a look.

LGTM; thank you. FYI, the e2e build is sometime flaky; especially when it fails at the Maven build stage, it is an indication of a transient error. I retried the last failed build for this PR and it succeeded.
